include/deal.II/matrix_free/tools.h

```diff
@@ -119,8 +146,381 @@ namespace MatrixFreeTools
+                     Number,
+                     VectorizedArrayType>
+          phi(matrix_free, dof_no, quad_no, first_selected_component);
+
+        const unsigned int n_lanes = VectorizedArrayType::size();
+
+        // local storage: buffer so that we access the global vector once
+        // note: may be larger then dofs_per_cell in the presence of
+        // constraints!
+        std::array<std::vector<Number>, n_lanes> diagonals_local_constrained;
+
+        for (unsigned int cell = range.first; cell < range.second; ++cell)
+          {
+            /*****************************************************************
+             * STEP 1: get relevant information from FEEvaluation
+             ****************************************************************/
+            const auto &       dof_info = matrix_free.get_dof_info(dof_no);
+            const unsigned int n_fe_components =
+              dof_info.start_components.back();
+            const unsigned int dofs_per_component = phi.dofs_per_component;
+
+            const unsigned int n_lanes_filled =
+              matrix_free.n_active_entries_per_cell_batch(cell);
+
+            std::array<const unsigned int *, n_lanes> dof_indices;
+            {
+              for (unsigned int v = 0; v < n_lanes_filled; ++v)
+                dof_indices[v] =
+                  dof_info.dof_indices.data() +
+                  dof_info
+                    .row_starts[(cell * n_lanes + v) * n_fe_components +
+                                first_selected_component]
+                    .first;
+
+              for (unsigned int v = n_lanes_filled; v < n_lanes; ++v)
+                dof_indices[v] = nullptr;
+            }
+
+            /*******************************************************************
+             * STEP 2: setup CSR storage of transposed locally-relevant
+             *   constraint matrix
+             ******************************************************************/
+            std::array<internal::LocalCSR<Number>, n_lanes> c_pools;
+
+            for (unsigned int v = 0; v < n_lanes_filled; ++v)
+              {
+                unsigned int index_indicators, next_index_indicators;
+
+                index_indicators =
+                  dof_info
+                    .row_starts[(cell * n_lanes + v) * n_fe_components +
+                                first_selected_component]
+                    .second;
+                next_index_indicators =
+                  dof_info
+                    .row_starts[(cell * n_lanes + v) * n_fe_components +
+                                first_selected_component + 1]
+                    .second;
+
+                /***************************************************************
+                 * STEP 2a: setup locally-relevant constraint matrix in a
+                 *   coordinate list (COO - see:
+                 *   https://en.wikipedia.org/wiki/Sparse_matrix#Coordinate_list_(COO))
```

Comment:
I do not think we need a link to COO here. One either understands coordinate list/COO or does not mess around in this code I believe.